velox/exec/fuzzer/ReferenceQueryRunner.h

```diff
@@ -32,6 +32,7 @@ class ReferenceQueryRunner {
   /// Executes SQL query returned by the 'toSql' method using 'input' data.
   /// Converts results using 'resultType' schema.
   virtual std::multiset<std::vector<velox::variant>> execute(
+      std::shared_ptr<memory::MemoryPool> rootPool,
```

Comment:
usually, we pass memory pool as raw pointer; it is fine to pass it as a reference as well, but let's not pass it as a shared_ptr

Why 'rootPool'? Can we pass a leaf pool and document that the method should use this pool to allocate memory for the result vector?

Reply:
Thanks for the review. I changed to pass raw pointer.
> Why 'rootPool'?

I notice we need leafPool and writerPool in PrestoQueryRunner.

In the SparkQueryRunner, a copy pool is also needed. During the copy, the pool of target vector should be different with that of source vector, otherwise the pool will be reused and copy does not happen. 
https://github.com/facebookincubator/velox/pull/10357/files#diff-04bd90dfc1fe1e50c098001f60ae4bdca4b61e662918661a141c7db948b28d47R273-R277

Do you think it makes sense to use rootPool? Thanks.

Reply:
@rui-mo Thank you for clarifying. It sounds like this API needs an "aggregate" pool to create a pool for writer, otherwise, it could use a "leaf" pool. I'm not sure I understand "otherwise the pool will be reused and copy does not happen". Would you elaborate a bit. A code reference above shows SparkQueryRunner::toSql method which doesn't seem to explain much.

If we pass an "aggregate" pool and use it to create a "leaf" pool, then we need to somehow ensure that "leaf" pool outlives the vectors that use it. How do we do that for the "result" vector?

Reply:
> I'm not sure I understand "otherwise the pool will be reused and copy does not happen". Would you elaborate a bit. A code reference above shows SparkQueryRunner::toSql method which doesn't seem to explain much.

Sorry for causing confusion, and this is the updated link [readArrowData](https://github.com/rui-mo/velox/blob/1d3eaaa2241d6e9402202be2de8ff55a9f0b7831/velox/functions/sparksql/fuzzer/SparkQueryRunner.cpp#L211-L215).

- Why we need this copy:

As below code shows, for string type `importFromArrowAsOwner` does not copy the buffer so the data is still maintained by Arrow after importing. A copy of is needed to manage the data in Velox.
https://github.com/facebookincubator/velox/blob/2fb811aa1a0761a1827702489f24032ba0583701/velox/vector/arrow/Bridge.cpp#L535-L537

- Why we need different pools for the source and target vectors:

As below code shows, the `set` method, which copies the buffer data, is only called if the pools are different.

https://github.com/facebookincubator/velox/blob/2fb811aa1a0761a1827702489f24032ba0583701/velox/vector/FlatVector-inl.h#L303

https://github.com/facebookincubator/velox/blob/2fb811aa1a0761a1827702489f24032ba0583701/velox/vector/FlatVector.cpp#L118

Reply:
> If we pass an "aggregate" pool and use it to create a "leaf" pool, then we need to somehow ensure that "leaf" pool outlives the vectors that use it. How do we do that for the "result" vector?

Thanks for catching this issue. I notice the fuzzers all provide leaf pool and aggregate pool, perhaps we need to pass both instead of a single aggregate pool. How do you think? Thanks.

Reply:
@rui-mo I'm wondering if we could pass an 'aggregate' pool to query runner's constructor and let it create child pools and store these in member variables. Would that work? (I'm out next week.)

Reply:
@mbasmanova Thanks for the suggestion. I assume it would work, and will try with that.